lib/apns/message_handler.ex

```diff
@@ -0,0 +1,82 @@
+defmodule APNS.MessageHandler do
+  require Logger
+
+  @payload_max_old 256
+  @payload_max_new 2048
+  @invalid_payload_size_code 7
+
+  def connect(%{config: config, ssl_opts: opts} = state, sender \\ APNS.Sender) do
+    sender.close(state.socket_apple)
+    host = to_char_list(config.apple_host)
+    port = config.apple_port
+
+    case sender.connect_socket(host, port, opts, config.timeout) do
+      {:ok, socket} -> {:ok, %{state | socket_apple: socket, counter: 0}}
+      {:error, reason} -> {:error, reason}
+    end
+  end
+
+  def push(_message, _state, sender \\ APNS.Sender)
+
+  def push(%APNS.Message{token: token} = message, state, _sender) when byte_size(token) != 64 do
+    APNS.Error.new(message.id, 5) |> state.config.callback_module.error()
+    state
+  end
+
+  def push(%APNS.Message{} = message, %{config: config, socket_apple: socket, queue: queue} = state, sender) do
+    limit = case message.support_old_ios do
+      nil -> config.payload_limit
+      true -> @payload_max_old
+      false -> @payload_max_new
+    end
+
+    case APNS.Payload.build_json(message, limit) do
+      {:error, :payload_size_exceeded} ->
+        APNS.Error.new(message.id, @invalid_payload_size_code) |> state.config.callback_module.error()
+        state
+
+      payload ->
+        binary_payload = APNS.Package.to_binary(message, payload)
+        case sender.send_package(socket, binary_payload) do
+          :ok ->
+            state = %{state | queue: [message | queue]}
+            Logger.debug("[APNS] success sending #{message.id} to #{message.token}")
+          {:error, reason} ->
+            state = %{state | queue: []}
+            Logger.error("[APNS] error (#{reason}) sending #{message.id} to #{message.token}")
+        end
+
+        if state.counter >= state.config.reconnect_after do
+          Logger.debug("[APNS] #{state.counter} messages sent, reconnecting")
+          connect(state, sender)
+        end
+
+        %{state | counter: state.counter + 1}
+    end
+  end
+
+  def handle_response(state, socket, data, worker_pid \\ self()) do
+    case <<state.buffer_apple :: binary, data :: binary>> do
+      <<8 :: 8, status :: 8, message_id :: integer-32, rest :: binary>> ->
+        APNS.Error.new(message_id, status) |> state.config.callback_module.error()
+
+        for message <- messages_after(state.queue, message_id) do
+          GenServer.cast(worker_pid, message)
```

Comment:
TL;DR: separate worker in to one for feedback and one for messages.

After discussing this more with @Linuus we've started to think that we're attacking this from the wrong angle. Before this PR both `MessageHandler`, `FeedbackHandler`, and `Worker` was in the same file. We wanted to protect against regressions and therefore added unit tests for this code. We didn't want to test all the logic through the GenServer interface so we basically just copy-pasted business logic into its own modules (`MessageHandler` and `FeedbackHandler`).

Perhaps this was wrong, what if we instead make two different Workers, one for Feedback and one for Message? Then we can inline the business logic in the Workers which would improve readability and not create falsey illusions that the Handlers are properly separated from the Worker.

This would also allow us to configure number of workers individually for Messages and Feedback. From what we understand now, if we want, say 100, worker because we want to send a lot of messages on separate connections we have the side-effect that we'll also call the Feedback-service 100 times each interval?

We haven't started on this yet and there might be some good arguments why this is a bad idea, so we're just putting it out there to hear what you think.